### ISSUE TITLE
Add helper hooks to common Python geo libraries (shapely, numpy, pyproj)

### ISSUE THREAD

**Issue by nyalldawson** [qgis/QGIS-Enhancement-Proposals#227]
# QGIS Enhancement: Add helper hooks to common Python geo libraries

**Date** 2021/05/26

**Author** Nyall Dawson (@nyalldawson)

**Contact** nyall dot dawson at gmail dot com

**maintainer** @nyalldawson

**Version** QGIS 3.X

# Summary

In recent years PyQGIS has made huge strides forward in usability and user-friendliness. This QEP seeks to extend this user-friendliness and convenience by adding in simple to use API to bridge PyQGIS classes to other non-QGIS spatial libraries such as shapely, pyproj and numpy.

This libraries are used heavily outside of QGIS, and providing user-friendly (i.e. one line) methods to convert between the equivalent QGIS object and the external library will greatly aid script development for developers.

**It is NOT proposed that these libraries become new dependencies for QGIS. (It will always be the recommended approach to use the QGIS native classes wherever possible in plugins in order to ensure that they run on all platforms which QGIS runs on without the need to install additional external dependencies.) Rather, the proposal describes a "soft" approach where the convenience methods are available whenever the library IS available to the Python environment only.**

This is the building block for a larger project which would see similar convenience methods exposed for conversions of QGIS layers to Pandas/Geopandas data frames (and some other exciting data science things which will be announced at a later stage!).

### General Approach

The general approach to be taken will be to add method "stubs" to the QGIS api, via c++ stubs. These methods will be non functional, and will immediately raise a QgsNotSupportedException exception.

For example:

```
class QgsGeometry
{
  ...
  #ifndef SIP_RUN
///@cond PRIVATE
  void as_shapely() const;
  static QgsGeometry from_shapely();
///@endcond
  #else
  /**
   * Returns the geometry as an equivalent shapely geometry (e.g. shape.Point, shapely.LineString, etc).
   *
   * Curved geometries types will be automatically segmentized in order to be representable as shapely geometries.
   * 
   * \throws QgsNotSupportedException if the shapely library is not available in the Python enviroment.
   */
  SIP_PYOBJECT as_shapely() const SIP_TYPEHINT( Optional[shapely.geometry] ) SIP_THROW( QgsNotSupportedException );
  % MethodCode
  sipCpp->as_shapely(); // throws QgsNotSupportedException
  % End

  /**
   * Converts a shapely geometry to the equivalent QGIS geometry.
   * \throws QgsNotSupportedException if the shapely library is not available in the Python enviroment.
   */
  static QgsGeometry from_shapely( SIP_PYOBJECT geometry SIP_TYPEHINT( shapely.geometry ) ) SIP_THROW( QgsNotSupportedException );
  % MethodCode
  sipCpp->from_shapely(); // throws QgsNotSupportedException
  % End

  #endif
};

void QgsGeometry::as_shapely() const
{
  throw QgsNotSupportedException( "The shapely library is not available in this Python environment");
}

QgsGeometry QgsGeometry::from_shapely()
{
  throw QgsNotSupportedException( "The shapely library is not available in this Python environment");
}
```

This will be accompanied by corresponding monkey patching in the python/core/__init__.py.in file:

```
try:
   import shapely
   
   def geometry_from_shapely(geometry: shapely.geometry) -> QgsGeometry:
       # conversion logic here
       return QgsGeometry( .. )

   def geometry_to_shapely(geometry: QgsGeometry) -> shapely.geometry:
       # conversion logic here
       return ...

   QgsGeometry.from_shapely = geometry_from_shapely
   QgsGeoemtry.as_shapely = geometry_to_shapely
       
except ModuleNotFoundError:
   pass
```

This approach has some key properties:
- The library is optional, and does not become a new dependency for QGIS
- User friendly exceptions are raised when the library is not available on a QGIS install, allowing the developer to handle this situation gracefully
- The convenience methods can be nicely documented and will be shown in the PyQGIS API documentation, along with all their correct type hints
- The type hints will be usable by IDEs, providing a super-nice development experience in PyCharm, etc

### Shapely

As described above, convenience methods will be added to convert to/from shapely geometries to corresponding QgsGeometry objects.

### PyProj

- Hooks will be added to QgsCoordinateReferenceSystem to allow conversion to/from PyProj CRS objects.

### numpy

- Convenience methods will be added to QgsRasterBlock to convert a QGIS raster data type to and from a numpy data type.
- A method will be added to QgsRasterBlock to return a numpy object (specifically a masked array), using the underlying data in the QgsRasterBlock without copying. E.g. something like:
```
def raster_block_as_numpy(self):
   src_array = numpy.frombuffer(self.data(), dtype=qgis_data_type_to_numeric_data_type(self.dataType()))
   return numpy.ma.masked_equal(src_array, self.noDataValue())

QgsRasterBlock.as_numpy = raster_block_as_numpy
```

Similar methods will be added for setting block data directly from a numpy array of the same data type.

High level api will be added to QgsRasterDataProvider and QgsRasterLayer to directly convert the whole raster to a numpy array using a similar approach.

## Performance Implications

As much of the "heavy lifting" will be done in c++ as possible, ensuring that the conversions are as optimised and performant as possible.

**Comment by m-kuhn** [comment]
Thanks for the proposal.

Where will the actual conversion functions / stub implementations live?

**Comment by nyalldawson** [comment]
@m-kuhn 

In python/core/__init__.py.in, with as many relevant helpers in C++ as possible. (In the case of these three libraries we've already got as much exposed through existing C++ as is needed for optimised conversion)

**Comment by m-kuhn** [comment]
If they are monkey patched in python, what are the C++ stubs needed for?

**Comment by nyalldawson** [comment]
Mostly so that the pyqgis docs are generated without issue 😁 (eg other monkey patched methods like the various processing ones aren't shown in the pyqgis docs)

**Comment by m-kuhn** [comment]
Are we not able to fix this on docs side? (CC @3nids )

**Comment by 3nids** [comment]
I gave it a try without success with processing in the past. I can have a look at it again.

**Comment by elpaso** [comment]
I like the idea, very useful indeed! I also have some concerns about the stubs if they are only for the PyDocs but if there isn't another way then it's ok.

**Comment by m-kuhn** [comment]
Would it be possible to shorten this to sip only code?

```
class QgsGeometry
{
  ...
  #ifdef SIP_RUN
  /**
   * Returns the geometry as an equivalent shapely geometry (e.g. shape.Point, shapely.LineString, etc).
   *
   * Curved geometries types will be automatically segmentized in order to be representable as shapely geometries.
   * 
   * \throws QgsNotSupportedException if the shapely library is not available in the Python enviroment.
   */
  SIP_PYOBJECT as_shapely() const SIP_TYPEHINT( Optional[shapely.geometry] ) SIP_THROW( QgsNotSupportedException );
  % MethodCode
  throw QgsNotSupportedException( "The shapely library is not available in this Python environment");
  % End
  #endif
};
```

**Comment by m-kuhn** [comment]
I forgot to say : +1 for the base idea of this proposal, it will be a very welcome addition for many.

**Comment by akadouri** [comment]
One thing to note is that Shapely is currently undergoing development to a 2.0 version https://github.com/Toblerity/Shapely/issues/962 which I believe changes it's geometry class (?). I don't know enough to know if this makes a difference, just mentioning it incase it does.

**Comment by martinfleis** [comment]
+1

Yes, @akadouri's point is correct and should be taken into account. While most of shapely's functionality will be backwards compatible, it may make more sense to build the conversion machinery on the new vectorised implementation @jorisvandenbossche is currently moving from pygeos to shapely.

As per geopandas, we're happy to help if you'll need something from our side.

**Comment by jorisvandenbossche** [comment]
Some random thoughts about the actual conversion logic (being unfamiliar with QGIS internals):

* The "easy" option is to use WKB, I assume, for converting between QGIS and Shapely geometries. I am wondering though, how does QGIS currently convert a `QgsGeometry` to a `GEOSGeometry`? (as QGIS already uses GEOS for some operations, I think). Looking at [qgsgeos.cpp](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsgeos.cpp), it seems there is a bunch of code actually constructing the coordinate sequences. I am not sure how easy it will be to do something similar in Python?

* The current sketch for `from_shapely` / `as_shapely` above seems to be about single geometries. In that case, the improvements on vectorized operations in pygeos (Shapely 2.0) are not that relevant here, as the scalar performance and API should be more or less the same. 
  It does me wonder, though, if it wouldn't be interesting to have similar conversion mechanisms for *arrays* of geometries (like all geometries that are in a layer in QGIS), from which you could construct for example a geopandas.GeoDataFrame without iterating through all scalar geometries in Python.

**Comment by nyalldawson** [comment]
@kannes that's really a "lowest common denominator". What I'm aiming for here is highly optimised, specialised code paths for integrating with these libraries.

**Comment by anitagraser** [comment]
@nyalldawson Could you share an update on how far you got with this idea and if/where it got stuck? I'm really interested in seeing this progress and wonder what is needed.

**Comment by nyalldawson** [comment]
This QEP is being archived in order to empty the issue tracker on this repository. Previous discussion and voting on the QEP remains valid and unchanged.
